File: signature/get_spectra.R
#!/usr/bin/Rscript

cat('Start at',date(),'\n')

args <- commandArgs(T)

myfun <- function(x){

path <- dirname(x)
sam <- basename(x)

setwd(path)

substitution <- read.table(sam,header=F,sep='\t',stringsAsFactors=F)
#colnames(substitution) <- c('chr','pos','ref','alt','flank5seq','flank3seq','gene','strand')
tmpname <- strsplit(sam,split='\\.')[[1]][1]
name <- paste(tmpname,'.ready.txt',sep='')

data1 <- data.frame('chr','pos','ref','alt','flank5seq','flank3seq','before','refcontext','after','altcontext','gene','strand')
write.table(data1,name,append=T,sep='\t',quote=F,row.names=F,col.names=F)

for (i in 1:nrow(substitution)) {
   tmp <- substitution[i,]
   chr=tmp[,1]
   pos=tmp[,2]
   ref=tmp[,3]
   alt=tmp[,4]
   flank5=tmp[,5]
   before=substr(flank5,10,10)
   flank3=tmp[,6]
   after=substr(flank3,1,1)
   gene=tmp[,7]
   strand=tmp[,8]
   if (strand==1) {
      #flank5=tmp[,5]
      #flank3=tmp[,6]
      if (ref=='C'|ref=='T') {
         refcontext=ref
         altcontext=alt

         rawdata=c(chr,pos,ref,alt,flank5,flank3,before,refcontext,after,altcontext,gene,strand)
      } 
      else {
         #reftmp=tmp[,3]
         if (ref=='A') {
            refcontext='T'}
         else if (ref=='G') {
            refcontext='C'}
         
         #alttmp=tmp[,4]
         if (alt=='A') {
            altcontext='T' }
         else if (alt=='C') {
            altcontext='G' }
         else if (alt=='G') {
            altcontext='C' }
         else if (alt=='T') {
            altcontext='A' }
         
        #betmp=substr(flank5,10,10)
        #if (betmp=='A') {
        #   before='T' }
        #else if (betmp=='C') { 
        #   before='G' }
        #else if (betmp=='G') {
        #   before='C' }
        #else if (betmp=='T') {
        #   before='A' }

        #aftmp=substr(flank3,1,1)
        #if (aftmp=='A') {
        #   after='T' }
        #else if (aftmp=='C') {
        #   after='G' }
        #else if (aftmp=='G') {
        #   after='C' }
        #else if (aftmp=='T') {
        #   after='A' }

        rawdata=c(chr,pos,ref,alt,flank5,flank3,before,refcontext,after,altcontext,gene,strand)
      }  
   }
   else {
      #flank5=tmp[,5]
      #flank3=tmp[,6]
      if (ref=='A'|ref=='G') {
         refcontext=ref
         altcontext=alt
        
         rawdata=c(chr,pos,ref,alt,flank5,flank3,before,refcontext,after,altcontext,gene,strand)
      }
      else {

         #reftmp=tmp[,3]
         if (ref=='C') {
            refcontext='G'}
         else if (ref=='T') {
            refcontext='A'}

         #alttmp=tmp[,4]
         if (alt=='A') {
            altcontext='T' }
         else if (alt=='C') {
            altcontext='G' }
         else if (alt=='G') {
            altcontext='C' }
         else if (alt=='T') {
            altcontext='A' }

         #betmp=substr(flank5,10,10)
         #if (betmp=='A') {
         #   before='T' }
         #else if (betmp=='C') {
         #   before='G' }
         #else if (betmp=='G') {
         #   before='C' }
         #else if (betmp=='T') {
         #   before='A' }

         #aftmp=substr(flank3,1,1)
         #if (aftmp=='A') {
         #   after='T' }
         #else if (aftmp=='C') {
         #   after='G' }
         #else if (aftmp=='G') {
         #   after='C' }
         #else if (aftmp=='T') {
         #   after='A' }

         rawdata=c(chr,pos,ref,alt,flank5,flank3,before,refcontext,after,altcontext,gene,strand)
      }
   }
data2 <- data.frame(chr=rawdata[1],pos=rawdata[2],ref=rawdata[3],alt=rawdata[4],flank5seq=rawdata[5],flank5seq=rawdata[6],before=rawdata[7],
refcontext=rawdata[8],after=rawdata[9],altcontext=rawdata[10],gene=rawdata[11],strand=rawdata[12])

write.table(data2,name,append=T,sep='\t',quote=F,row.names=F,col.names=F)
}

}

#files <- list.files(paste('/online/home/chenyl/dongfang/',pro,'/',group,sep=''),'ready.txt')
#for (x in files){
#  myfun(x)
#}

myfun(args[1])

cat('Ended at',date(),'\n')

File: signature/get_spectra.py
#!/usr/bin/env python3

"""This script is used to get variation context, Used to replace get_spectra.R, 
   because R script run slowly,which need ten minutes per sample
"""

__author__ = "yulong chen"
__email__ = "chenyl@cloudhealth99.com"

import os
import sys
import time

start=time.clock()

# Check argv and folder existance
if not len(sys.argv) == 2: 
    sys.exit('Usage: python3 %s file' % sys.argv[0])
if not os.path.isfile(sys.argv[1]): 
    sys.exit("%s not exists." % sys.argv[1])

dir = os.path.dirname(sys.argv[1])
os.chdir(dir)
file = os.path.basename(sys.argv[1])
outfile = open(file.replace('.tmp.txt','.ready.txt'),'w')
outfile.write('chr'+'\t'+'pos'+'\t'+'ref'+'\t'+'alt'+'\t'+'flank5seq'+'\t'+'flank3seq'+'\t'+'before' \
              +'\t'+'refcontext'+'\t'+'after'+'\t'+'altcontext'+'\t'+'gene'+'\t'+'strand'+'\n')

##('chr','pos','ref','alt','flank5seq','flank3seq','before','refcontext','after','altcontext','gene','strand')
dict={'A':'T','T':'A','C':'G','G':'C'}

with open(file,'r') as f:
    for line in f:
        if line.startswith('#'):
            continue
        chr,pos,ref,alt,flank5seq,flank3seq,gene,strand=line.strip().split('\t')
        before=flank5seq[9]
        after=flank3seq[0]
        if strand == '1':
            refcontext=ref
            altcontext=alt
        else:
            refcontext=dict[ref]
            altcontext=dict[alt]

        outfile.write(chr+'\t'+pos+'\t'+ref+'\t'+alt+'\t'+flank5seq+'\t'+flank3seq+'\t'+before \
                              +'\t'+refcontext+'\t'+after+'\t'+altcontext+'\t'+gene+'\t'+strand+'\n')
outfile.close()
end=time.clock()
print("Time used:",start-end)
